runtime/src/iree/io/formats/gguf/gguf_format.c

```diff
@@ -0,0 +1,641 @@
+  IREE_RETURN_IF_ERROR(
+      iree_io_gguf_parse_uint64(&contents, &metadata_kv_count));
+
+  // If there are no tensors then no-op the parse. Probably not what the user
+  // wanted but it's legal.
+  if (tensor_count == 0) return iree_ok_status();
+
+  iree_io_gguf_parser_t parser = {
+      .file_handle = file_handle,
+      .index = index,
+      .alignment = GGUF_DEFAULT_ALIGNMENT,  // may be overridden
+      .tensor_data_offset = 0,              // to be calculated
+      .tensor_data_size = 0,                // to be calculated
+  };
+
+  // Scope data to the remainder of the file and enumerate all metadata pairs.
+  // Upon return the contents will start immediately after the header and at
+  // the start of tensor info.
+  IREE_RETURN_IF_ERROR(iree_io_gguf_enumerate_metadata_kv(
+      &contents, metadata_kv_count, iree_io_gguf_parse_metadata, &parser));
+
+  // Scan forward through the tensor info to find where the tensor data base
+  // offset is in the file. Unfortunately GGUF was designed without this offset
+  // and because tensor info is variable length we cannot determine absolute
+  // file offsets without doing two scans.
+  iree_const_byte_span_t tensor_info_contents = contents;
+  IREE_RETURN_IF_ERROR(iree_io_gguf_enumerate_tensor_info(
+      &tensor_info_contents, tensor_count, NULL, &parser));
+
+  // Calculate where the tensor data begins in the file. This respects the
+  // default alignment or the general.alignment specified by the file.
+  parser.tensor_data_offset = iree_align_uint64(
+      (uint64_t)(tensor_info_contents.data - file_contents.data),
+      parser.alignment);
+  parser.tensor_data_size =
+      file_contents.data_length - parser.tensor_data_offset;
+
+  // Scan forward through the tensor info now that we know the tensor data
+  // offset and add the tensor entries.
+  IREE_RETURN_IF_ERROR(iree_io_gguf_enumerate_tensor_info(
+      &contents, tensor_count, iree_io_gguf_append_tensor_info, &parser));
+
+  return iree_ok_status();
+}
+
+IREE_API_EXPORT iree_status_t iree_io_parse_gguf_index(
+    iree_io_file_handle_t* file_handle, iree_io_parameter_index_t* index) {
+  IREE_ASSERT_ARGUMENT(index);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  // Today we only support memory files.
+  // TODO(benvanik): support iree_io_stream_t wrapping for parsing the index.
+  if (iree_io_file_handle_type(file_handle) !=
+      IREE_IO_FILE_HANDLE_TYPE_HOST_ALLOCATION) {
+    IREE_TRACE_ZONE_END(z0);
+    return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
+                            "non-memory gguf files not yet supported");
+  }
+  iree_byte_span_t host_allocation =
+      iree_io_file_handle_primitive(file_handle).value.host_allocation;
+
+  iree_status_t status = iree_io_parse_gguf_index_from_memory(
+      file_handle,
+      iree_make_const_byte_span(host_allocation.data,
+                                host_allocation.data_length),
+      index);
+
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
```

Comment:
Ugh. This is... not a good format.

Reply:
Yeah - does a few things right but needs a few more iterations. I'm not sure they have the right pressure to make it better, though, and I suspect it'll just get more weird stuff shoved into it. That you need the full format to parse the file but their own code #ifdefs some of the info based on compiler flags shows me they haven't really worked through compatibility. At least it has alignment :P